common/message_algos/src/filters/message_filter.rs

```diff
@@ -0,0 +1,14 @@
+pub use crate::Envelope;
+
+/// Stateful message filter that accepts or rejects messages based on some criteria.
+pub trait MessageFilter: Send {
+    type Message: Send + Clone;
+
+    fn filter(&mut self, _message: &Envelope<Self::Message>) -> FilterDecision;
+}
```

Comment:
Ok for the trait. However, I miss an implementation for the filter that matters the most (from my point of view): 

* discarding messages that arrive too late i.e with an event timestamp not below the system time minus an accepted delay.

Reply:
You mean sth rougly like this?

```rust
struct TooLateFilter {
    clock: clock::Clock,
    delta: Duration,
}

impl MessageFilter for TooLateFilter {
   type Message = CollectdMessage;

   fn filter(&mut self, envelope: &Envelope<Self::Message>) -> FilterDecision {
     let age = self.clock.now() - envelope.message.timestamp(); // XXX
     if age > self.delta {
        FilterDecision::Reject
     } else {
       FilterDecision::Accept
    }
   }
}
```

Reply:
Yes but using `envelope.received_at` rather than an additional clock.

```
   fn filter(&mut self, envelope: &Envelope<Self::Message>) -> FilterDecision {
     let delay= envelope.received_at - envelope.message.timestamp()
     if delay> self.delta {
        FilterDecision::Reject
     } else {
       FilterDecision::Accept
    }
   }
```